

# Project: Medical Appointment No-Shows Investigation

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

The dataset object of this investigation is a series of appointments records at the Brazilian Unified Health System (SUS). The SUS is a free and public healthcare system based in equality and universal access for the people of Brazil. In 2019 17.3 millions of brazilians looked for some primary attetion service at the SUS in the last six months before particiapation in the [PNS-2019 assessment](https://agenciadenoticias.ibge.gov.br/agencia-sala-de-imprensa/2013-agencia-de-noticias/releases/29203-pns-2019-quem-mais-utiliza-o-sus-avaliou-mais-positivamente-a-qualidade-dos-servicos-de-atencao-primaria-a-saude). (source:Ibge) 

The SUS system is characterized by the tranditional limitation of public systems, the high demands rates, high no-show rates and long wating time for appoinments. 

The primary posed question for this dataset is what factors or features of an appoiment can drive to a no-show event. Is there any correlation between, gender, age or schedule/appoinment caracteristcs (day of the week, time of the day,  month, season, timespan,. etc..)  that are important for a no-show event?




Dataset description: 

>This dataset collects information from 100k medical appointments in Brazil and is focused on the question of whether or not patients show up for their appointment. A number of characteristics about the patient are included in each row.
>‘ScheduledDay’ tells us on what day the patient set up their appointment.
‘Neighborhood’ indicates the location of the hospital.
‘Scholarship’ indicates whether or not the patient is enrolled in Brasilian welfare program Bolsa Família.
>Be careful about the encoding of the last column: it says ‘No’ if the patient showed up to their appointment, and ‘Yes’ if they did not show up.


**Delete me !!!**
> **Tip**: In this section of the report, provide a brief introduction to the dataset you've selected for analysis. At the end of this section, describe the questions that you plan on exploring over the course of the report. Try to build your report around the analysis of at least one dependent variable and three independent variables.
>
> If you haven't yet selected and downloaded your data, make sure you do that first before coming back here. If you're not sure what questions to ask right now, then make sure you familiarize yourself with the variables and the dataset context for ideas of what to explore.

In [1]:
# Use this cell to set up import statements for all of the packages that you
#   plan to use.
# Remember to include a 'magic word' so that your visualizations are plotted
#   inline with the notebook. See this page for more:
#   http://ipython.readthedocs.io/en/stable/interactive/magics.html

In [2]:
# Importing pandas for data manipulation
# Importing matplolib ans seaborn for data visualization 
# Import numpy for numerical operations

import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sbn
import numpy as np

<a id='wrangling'></a>
## Data Wrangling



### General Properties

In [3]:
df = pd.read_csv('noshowappointments-kagglev2-may-2016.csv')

In [4]:
# Looking at dataset head for data import verification 
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

#### Missing values and data types 
*using info for missing data and data types verification*

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   PatientId       110527 non-null  float64
 1   AppointmentID   110527 non-null  int64  
 2   Gender          110527 non-null  object 
 3   ScheduledDay    110527 non-null  object 
 4   AppointmentDay  110527 non-null  object 
 5   Age             110527 non-null  int64  
 6   Neighbourhood   110527 non-null  object 
 7   Scholarship     110527 non-null  int64  
 8   Hipertension    110527 non-null  int64  
 9   Diabetes        110527 non-null  int64  
 10  Alcoholism      110527 non-null  int64  
 11  Handcap         110527 non-null  int64  
 12  SMS_received    110527 non-null  int64  
 13  No-show         110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


***using describe for visualizing statistics and ranges***

In [6]:
df.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.921784e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172614e+12   5.640286e+06      18.000000       0.000000   
50%    3.173184e+13   5.680573e+06      37.000000       0.000000   
75%    9.439172e+13   5.725524e+06      55.000000       0.000000   
max    9.999816e+14   5.790484e+06     115.000000       1.000000   

        Hipertension       Diabetes     Alcoholism        Handcap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

        SMS_received  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [7]:
## Numeber of unique values
df.nunique()

PatientId          62299
AppointmentID     110527
Gender                 2
ScheduledDay      103549
AppointmentDay        27
Age                  104
Neighbourhood         81
Scholarship            2
Hipertension           2
Diabetes               2
Alcoholism             2
Handcap                5
SMS_received           2
No-show                2
dtype: int64


### Data Cleaning:

**Findings:**

1. General:
    - Dataset with 110527 rows and 14 columns with types:float64(1), int64(8), object(5)
    - Upper and lower caps mixed in column names
    - No missing values found
    - "-" charcter at column name
1. Column PatientId - type float64. It makes more sense to have this as strings. 
1. Column AppoitentID - type int64. It makes more sense to have this as strings
1. Colunn SchedulleDay and AppointmentDay make sense for analisys in datetime format.
1. There are multiple appoimments for the same patient identification, not really a problem, but require attention not to bias the analisys
1. Need fix Age of value or values == -1 

    

***Fixing column names seting lower and removing special chars***

In [8]:
columns = df.columns
columns = columns.str.lower()
columns = columns.str.replace('-','')
columns = columns.str.replace('_','')
df.columns = columns


***Checking fix.***

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   patientid       110527 non-null  float64
 1   appointmentid   110527 non-null  int64  
 2   gender          110527 non-null  object 
 3   scheduledday    110527 non-null  object 
 4   appointmentday  110527 non-null  object 
 5   age             110527 non-null  int64  
 6   neighbourhood   110527 non-null  object 
 7   scholarship     110527 non-null  int64  
 8   hipertension    110527 non-null  int64  
 9   diabetes        110527 non-null  int64  
 10  alcoholism      110527 non-null  int64  
 11  handcap         110527 non-null  int64  
 12  smsreceived     110527 non-null  int64  
 13  noshow          110527 non-null  object 
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


***Changing:***
- *patientid and appointimentid types to string.* 

- *scheduleday and appoinmentday to datetime*


In [10]:
df.patientid = df.patientid.astype(int).astype(str)
df.appointmentid = df.appointmentid.astype(str)
df.scheduledday = pd.to_datetime(df.scheduledday)
df.appointmentday = pd.to_datetime(df.appointmentday)
#verifing
df.head(1)

patientid appointmentid gender              scheduledday  \
0  29872499824296       5642903      F 2016-04-29 18:38:08+00:00   

             appointmentday  age    neighbourhood  scholarship  hipertension  \
0 2016-04-29 00:00:00+00:00   62  JARDIM DA PENHA            0             1   

   diabetes  alcoholism  handcap  smsreceived noshow  
0         0           0        0            0     No

In [11]:
df.describe()

age    scholarship   hipertension       diabetes  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean       37.088874       0.098266       0.197246       0.071865   
std        23.110205       0.297675       0.397921       0.258265   
min        -1.000000       0.000000       0.000000       0.000000   
25%        18.000000       0.000000       0.000000       0.000000   
50%        37.000000       0.000000       0.000000       0.000000   
75%        55.000000       0.000000       0.000000       0.000000   
max       115.000000       1.000000       1.000000       1.000000   

          alcoholism        handcap    smsreceived  
count  110527.000000  110527.000000  110527.000000  
mean        0.030400       0.022248       0.321026  
std         0.171686       0.161543       0.466873  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000  
50%         0.000000       0.000000       0.000000  
75%         0.000000       0.000000       1.000000  
max         1.000000       4.000000       1.000000

In [12]:
# Checking unique lists for mistakes, typos, etc. 
print(sorted(df.neighbourhood.unique()))
print(sorted(df.handcap.unique()))

['AEROPORTO', 'ANDORINHAS', 'ANTÔNIO HONÓRIO', 'ARIOVALDO FAVALESSA', 'BARRO VERMELHO', 'BELA VISTA', 'BENTO FERREIRA', 'BOA VISTA', 'BONFIM', 'CARATOÍRA', 'CENTRO', 'COMDUSA', 'CONQUISTA', 'CONSOLAÇÃO', 'CRUZAMENTO', 'DA PENHA', 'DE LOURDES', 'DO CABRAL', 'DO MOSCOSO', 'DO QUADRO', 'ENSEADA DO SUÁ', 'ESTRELINHA', 'FONTE GRANDE', 'FORTE SÃO JOÃO', 'FRADINHOS', 'GOIABEIRAS', 'GRANDE VITÓRIA', 'GURIGICA', 'HORTO', 'ILHA DAS CAIEIRAS', 'ILHA DE SANTA MARIA', 'ILHA DO BOI', 'ILHA DO FRADE', 'ILHA DO PRÍNCIPE', 'ILHAS OCEÂNICAS DE TRINDADE', 'INHANGUETÁ', 'ITARARÉ', 'JABOUR', 'JARDIM CAMBURI', 'JARDIM DA PENHA', 'JESUS DE NAZARETH', 'JOANA D´ARC', 'JUCUTUQUARA', 'MARIA ORTIZ', 'MARUÍPE', 'MATA DA PRAIA', 'MONTE BELO', 'MORADA DE CAMBURI', 'MÁRIO CYPRESTE', 'NAZARETH', 'NOVA PALESTINA', 'PARQUE INDUSTRIAL', 'PARQUE MOSCOSO', 'PIEDADE', 'PONTAL DE CAMBURI', 'PRAIA DO CANTO', 'PRAIA DO SUÁ', 'REDENÇÃO', 'REPÚBLICA', 'RESISTÊNCIA', 'ROMÃO', 'SANTA CECÍLIA', 'SANTA CLARA', 'SANTA HELENA', 'SANTA

In [13]:
# Checking Age column for typo error

print('Age min value:{} , Age Max value:{}'.format(df.age.min(),df.age.max()))


Age min value:-1 , Age Max value:115


In [14]:

df.drop(df[df.age == -1].index, axis=0,inplace=True)

#confirm drop outlier

print('Age min value:{} , Age Max value:{}'.format(df.age.min(),df.age.max()))


Age min value:0 , Age Max value:115


***One possible influece to noshow is the time between scheduled and the appoinment day.***

*** Creating the column timespan column as difference in days between appointmentday and scheduledday colouns. ***

In [27]:
df['timespan']= df['appointmentday'].dt.date - df['scheduledday'].dt.date

In [28]:
df.head(50)

patientid appointmentid gender              scheduledday  \
0    29872499824296       5642903      F 2016-04-29 18:38:08+00:00   
1   558997776694438       5642503      M 2016-04-29 16:08:27+00:00   
2     4262962299951       5642549      F 2016-04-29 16:19:04+00:00   
3      867951213174       5642828      F 2016-04-29 17:29:31+00:00   
4     8841186448183       5642494      F 2016-04-29 16:07:23+00:00   
5    95985133231274       5626772      F 2016-04-27 08:36:51+00:00   
6   733688164476661       5630279      F 2016-04-27 15:05:12+00:00   
7     3449833394123       5630575      F 2016-04-27 15:39:58+00:00   
8    56394729949972       5638447      F 2016-04-29 08:02:16+00:00   
9    78124564369297       5629123      F 2016-04-27 12:48:25+00:00   
10  734536231958495       5630213      F 2016-04-27 14:58:11+00:00   
11    7542951368435       5620163      M 2016-04-26 08:44:12+00:00   
12  566654781423437       5634718      F 2016-04-28 11:33:51+00:00   
13  911394617215919       5636249      M 2016-04-28 14:52:07+00:00   
14   99884723334928       5633951      F 2016-04-28 10:06:24+00:00   
15      99948393975       5620206      F 2016-04-26 08:47:27+00:00   
16   84574392942817       5633121      M 2016-04-28 08:51:47+00:00   
17   14794966191172       5633460      F 2016-04-28 09:28:57+00:00   
18   17135378245248       5621836      F 2016-04-26 10:54:18+00:00   
19    7223289184215       5640433      F 2016-04-29 10:43:14+00:00   
20  622257462899397       5626083      F 2016-04-27 07:51:14+00:00   
21   12154843752835       5628338      F 2016-04-27 10:50:45+00:00   
22  863229818887631       5616091      M 2016-04-25 13:29:16+00:00   
23  213753979425692       5634142      F 2016-04-28 10:27:05+00:00   
24    8734857996885       5641780      F 2016-04-29 14:19:19+00:00   
25    5819369978796       5624020      M 2016-04-26 15:04:17+00:00   
26      25787851512       5641781      F 2016-04-29 14:19:42+00:00   
27   12154843752835       5628345      F 2016-04-27 10:51:45+00:00   
28    5926171692527       5642400      M 2016-04-29 15:48:02+00:00   
29    1225776163665       5642186      F 2016-04-29 15:16:29+00:00   
30     342815551642       5628068      F 2016-04-27 10:24:52+00:00   
31     311284853849       5628907      M 2016-04-27 12:07:14+00:00   
32   52883562769472       5637908      M 2016-04-29 07:30:27+00:00   
33    7653516999712       5616921      F 2016-04-25 15:01:04+00:00   
34   19999756532994       5637963      F 2016-04-29 07:34:30+00:00   
35   78162635161611       5637968      M 2016-04-29 07:34:42+00:00   
36   72984587621439       5637975      F 2016-04-29 07:35:19+00:00   
37    1578131861739       5637986      F 2016-04-29 07:35:56+00:00   
38    5873315843778       5609446      M 2016-04-20 15:54:18+00:00   
39   14556231793236       5639644      F 2016-04-29 09:21:13+00:00   
40  996868412638744       5635881      F 2016-04-28 14:14:16+00:00   
41  822432466381793       5633339      F 2016-04-28 09:20:36+00:00   
42   25965426543339       5632906      F 2016-04-28 08:34:54+00:00   
43   71558956872246       5641620      F 2016-04-29 14:02:16+00:00   
44     274164858852       5635414      F 2016-04-28 13:27:27+00:00   
45    4982378572899       5635842      F 2016-04-28 14:11:24+00:00   
46     137943696338       5615608      M 2016-04-25 12:44:36+00:00   
47     589458459955       5633116      F 2016-04-28 08:51:26+00:00   
48    8545415176986       5618643      F 2016-04-26 07:19:49+00:00   
49   92235587471561       5534656      F 2016-03-31 17:11:17+00:00   

              appointmentday  age      neighbourhood  scholarship  \
0  2016-04-29 00:00:00+00:00   62    JARDIM DA PENHA            0   
1  2016-04-29 00:00:00+00:00   56    JARDIM DA PENHA            0   
2  2016-04-29 00:00:00+00:00   62      MATA DA PRAIA            0   
3  2016-04-29 00:00:00+00:00    8  PONTAL DE CAMBURI            0   
4  2016-04-29 00:00:00+00:00   56    JARDIM DA PENHA            0   
5  2016-04-29 00:00:00+00

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. Compute statistics and create visualizations with the goal of addressing the research questions that you posed in the Introduction section. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.



In [29]:
## masking

noshow = df.noshow == 'Yes'
show  = df.noshow == 'No'


### Research Question 1 :   *Is there a prefered gender for appointment no-show?*

In [ ]:
df[noshow].

### Research Question 2  (Replace this header name!)

In [ ]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work, you should save a copy of the report in HTML or PDF form via the **File** > **Download as** submenu. Before exporting your report, check over it to make sure that the flow of the report is complete. You should probably remove all of the "Tip" quotes like this one so that the presentation is as tidy as possible. Congratulations!

References:
    
[1]    https://agenciadenoticias.ibge.gov.br/agencia-sala-de-imprensa/2013-agencia-de-noticias/releases/29203-pns-2019-quem-mais-utiliza-o-sus-avaliou-mais-positivamente-a-qualidade-dos-servicos-de-atencao-primaria-a-saude
    